File: OJ/leet442/leet442.py
# leet442

from typing import List


class Solution:
    def findDuplicates1(self, A: List[int]) -> List[int]:
        for i in range(len(A)):
            while A[A[i]-1] != A[i]:
                A[A[i]-1], A[i] = A[i], A[A[i]-1] 
        return [ x for i, x in enumerate(A) if x != i+1 ] 
    
    def findDuplicates(self, A: List[int]) -> List[int]:
        ans = []
        for x in A:
            i = abs(x) - 1
            if A[i] > 0:
                A[i] *= -1
            else:
                ans.append(abs(x))
        return ans


print(Solution().findDuplicates([4,3,2,7,8,2,3,1]))

File: OJ/leet442/leet442.cpp
/*
leet442


*/

#include "../utils/LeetCpp.utils.hpp"

using namespace std;

#define ALL(A) A.begin(),A.end()
using VI=vector<int>;
using VVI=vector<VI>;


class Solution {
public:
    vector<int> findDuplicates(vector<int>& A) {
        vector<int> ans;
        for (auto x: A) {
            x = abs(x);
            if ( A[x - 1] < 0) ans.push_back( x );
            else A[x - 1] *= -1;
        }
        return ans;
    }
};


int main(int argc, char const *argv[])
{
    Solution sol;   
    {
        VI A {4,3,2,7,8,2,3,1}; 
        cout << sol.findDuplicates(A) << endl;
    }
    {
        VI A {1,1,2}; 
        cout << sol.findDuplicates(A) << endl;
    }
    {
        VI A {1}; 
        cout << sol.findDuplicates(A) << endl;
    }
    return 0;
}
